### PR TITLE
Restore feature: Auto-select backup set for PITR

### PR DESCRIPTION
**Pull request by cmwshang**
Reference: https://github.com/pgbackrest/pgbackrest/projects/2#card-8956672

When performing a time-based restore (PITR), if a backup set is not specified using option `--set`, then the behavior of the restore was to choose the latest backup as the starting point; in some instances this was passed the target time which then did not have the desired result.

This feature will attempt to select an appropriate backup set based on the provided time when the `--set` option is not provided. The caveat is that the time must conform to the following formatting rules:

YYYY-MM-DD HH:MM:SS with optional msec (up to 6 digits and separated from minutes by a comma or period), optional timezone offset +/- HH or HHMM or HH:MM, where offset boundaries are UTC-12 to UTC+14. If the timezone is omitted, then the time provided is assumed to be in the local timezone of the system.

If any other time format is provided, the system will default to using the latest backup if the `--set` option is not provided.

### CODE REVIEW

File: doc/xml/user-guide.xml [hunk 1]
```diff
@@ -2006,10 +2011,10 @@
             </execute>
         </execute-list>
 
-        <p>The default behavior for restore is to use the last backup but an earlier backup can be specified with the <br-option>{[dash]}-set</br-option> option.</p>
+        <p>The default behavior for time-based restore, if the <br-option>{[dash]}-set</br-option> option is not set, is to attempt to discover an earlier backup to play forward from; if a backup set cannot be found, then restore will default to the latest backup which, as shown earlier, may not give the desired result.</p>
 
         <execute-list host="{[host-pg1]}">
-            <title>Stop <postgres/>, restore from the selected backup, and start <postgres/></title>
+            <title>Stop <postgres/>, allow <backrest/> to discover the backup, and start <postgres/></title>
```
Comment (dwsteele) [inline doc/xml/user-guide.xml]: This title should still mention that a restore is happening, maybe "restore from auto-selected backup" or "restore from discovered backup".  Maybe auto-select is generally a better term than discover?

File: src/command/restore/restore.c [hunk 2]
```diff
@@ -106,6 +108,94 @@ restorePathValidate(void)
...
+
+    ASSERT(targetTime != NULL);
+
+    time_t result = 0;
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // Build the regex to accept formats: YYYY-MM-DD HH:MM:SS with optional msec (up to 6 digits and separated from minutes by
+        // a comma or period), optional timezone offset +/- HH or HHMM or HH:MM, where offset boundaries are UTC-12 to UTC+14
+        String *expression = strNew(
```
[the remark this answers is not in the document]
Comment (cmwshang) [reply src/command/restore/restore.c]: ok

File: src/command/restore/restore.c [hunk 3]
```diff
@@ -106,6 +108,94 @@ restorePathValidate(void)
...
+            "^(2[0-9]{3})-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01]) (0[0-9]|1[0-9]|2[0-3]):(0[0-9]|[1-5][0-9]):(0[0-9]|[1-5][0-9])"
+            "(\\,[0-9]{1,6}|\\.[0-9]{1,6})?(((\\+(0[0-9]|1[0-3])(:?(00|30|45))?)|(\\+(14)(:?00)?))|((\\-(0[0-9]|1[0-1])"
+            "(:?(00|30|45))?)|(\\-12)(:?00)?))?$");
```
[the remark this answers is not in the document]
Comment (cmwshang) [reply src/command/restore/restore.c]: It is not wrong, but now that we have the date/time validation functions, I agree that we can simplify the regex and then use the functions to validate.

File: src/command/restore/restore.c [hunk 4]
```diff
@@ -106,6 +108,94 @@ restorePathValidate(void)
...
+            int dtSecond = cvtZToInt(strPtr(strSubN(datetime, 17, 2)));
+
+            String *timeTargetZone = strSub(targetTime, 19);
+
+            // Determine if the remainder contains a timezone offset - if not, then local time is assumed
+            int idxPlus = strChr(timeTargetZone, '+');
+            int idxMinus = strChr(timeTargetZone, '-');
+            if (idxPlus != -1 || idxMinus != -1)
+            {
+                String *timezoneOffset = strSub(timeTargetZone, (size_t)(idxPlus == -1 ? idxMinus : idxPlus));
```
[the remark this answers is not in the document]
Comment (cmwshang) [reply src/command/restore/restore.c]: There may be milliseconds before the timezone, so the +/- may not be in index 0. I will try to make the comment clearer.

File: src/command/restore/restore.c [hunk 5]
```diff
@@ -106,6 +108,94 @@ restorePathValidate(void)
...
+                // Include the sign with the hour
+                int tzHour = cvtZToInt(strPtr(strSubN(timezoneOffset, 0, 3)));
+                int tzMinute = 0;
+
+                // If minutes are included in timezone offset then see if separated by a colon or not and extract accordingly
+                if (strSize(timezoneOffset) > 3)
+                {
+                    int colonIdx = strChr(timezoneOffset, ':');
+                    if (colonIdx != -1)
+                        tzMinute = cvtZToInt(strPtr(strSubN(timezoneOffset, (size_t)colonIdx+1, 2)));
```
[the remark this answers is not in the document]
Comment (cmwshang) [reply src/command/restore/restore.c]: Agreed

File: src/command/restore/restore.c [hunk 6]
```diff
@@ -130,7 +220,41 @@ restoreBackupSet(InfoBackup *infoBackup)
             if (infoBackupDataTotal(infoBackup) == 0)
                 THROW(BackupSetInvalidError, "no backup sets to restore");
 
-            backupSet = infoBackupData(infoBackup, infoBackupDataTotal(infoBackup) - 1).backupLabel;
+            time_t timeTargetEpoch = 0;
+
+            // If the recovery type is type, attempt to determine the backup set
```
[the remark this answers is not in the document]
Comment (cmwshang) [reply src/command/restore/restore.c]: Fixed

File: doc/xml/user-guide.xml (hunk 1 above)
Comment (cmwshang) [reply doc/xml/user-guide.xml]: Sounds good

File: src/command/restore/restore.c [hunk 7]
```diff
@@ -106,6 +108,94 @@ restorePathValidate(void)
...
+                result = mktime(
+                    &(struct tm){.tm_sec = dtSecond, .tm_min = dtMinute, .tm_hour = dtHour, .tm_mday = dtDay, .tm_mon = dtMonth - 1,
+                    .tm_year = dtYear - 1900, .tm_isdst = -1});
+            }
+        }
+        else
+        {
+            LOG_WARN_FMT(
+                "automatic backup set selection cannot be performed with provided time format '%s', latest backup set will be used"
+                "\nHINT: time format must be YYYY-MM-DD HH:MM:SS with optional msec and optional timezone (+/- HH or HHMM or HH:MM)"
```
[the remark this answers is not in the document]
Comment (cmwshang) [reply src/command/restore/restore.c]: The first "format" is what I believe you meant and I agree.

File: src/command/restore/restore.c [hunk 8]
```diff
@@ -130,7 +220,41 @@ restoreBackupSet(InfoBackup *infoBackup)
...
+                    if (backupSet == NULL)
+                    {
+                        LOG_WARN_FMT(
+                            "unable to find backup set with stop time less than '%s', latest backup set will be used",
+                            strPtr(cfgOptionStr(cfgOptTarget)));
+                    }
+                }
+            }
+
+            if (backupSet == NULL)
```
[the remark this answers is not in the document]
Comment (cmwshang) [reply src/command/restore/restore.c]: agreed

**Comment by dwsteele** [comment]
Looks good!  Committed at dbaa5e3, f46d1fa74c4c0, 856980ae99e.

**Pull request closed without merging**